File: examples/secondary_fota/src/ota_service.c
/*
* Copyright (C) INGCHIPS. All rights reserved.
* This code is INGCHIPS proprietary and confidential.
* Any use of the code for whatever purpose is subject to
* specific written permission of INGCHIPS.
*/

#include <stdio.h>
#include <string.h>
#include "ingsoc.h"

#include "att_db.h"
#include "ota_service.h"
#include "platform_api.h"
#include "rom_tools.h"

#define PAGE_SIZE (8192)

extern ota_ver_t this_version;

#include "../data/gatt.const"

#define ATT_OTA_HANDLE_VER          HANDLE_FOTA_VERSION
#define ATT_OTA_HANDLE_DATA         HANDLE_FOTA_DATA
#define ATT_OTA_HANDLE_CTRL         HANDLE_FOTA_CONTROL

static uint8_t  ota_ctrl[] = {OTA_STATUS_DISABLED};
static uint8_t  ota_downloading = 0;
static uint32_t ota_start_addr = 0;
static uint32_t ota_page_offset = 0;
static uint8_t  page_buffer[PAGE_SIZE];

int ota_write_callback(uint16_t att_handle, uint16_t transaction_mode, uint16_t offset, const uint8_t *buffer, uint16_t buffer_size)
{
    if (transaction_mode != ATT_TRANSACTION_MODE_NONE)
    {
        platform_printf("transaction_mode: %d\n", transaction_mode);
        return 0;
    }

    if (att_handle == ATT_OTA_HANDLE_CTRL)
    {
        if (OTA_CTRL_START == buffer[0])
        {
            ota_ctrl[0] = OTA_STATUS_OK;
            ota_start_addr = 0;
            ota_downloading = 0;
            return 0;
        }

        switch (buffer[0])
        {
        case OTA_CTRL_PAGE_BEGIN:
            ota_start_addr = *(uint32_t *)(buffer + 1);
            if (ota_start_addr & 0x3)
            {
                ota_ctrl[0] = OTA_STATUS_ERROR;
                return 0;
            }
            else
                ota_ctrl[0] = OTA_STATUS_OK;
            ota_downloading = 1;
            ota_page_offset = 0;
            break;
        case OTA_CTRL_PAGE_END:
            program_flash(ota_start_addr, page_buffer, ota_page_offset);

            ota_downloading = 0;
            {
                uint16_t len = *(uint16_t *)(buffer + 1);
                uint16_t crc_value = *(uint16_t *)(buffer + 3);
                if (ota_page_offset < len)
                {
                    ota_ctrl[0] = OTA_STATUS_WAIT_DATA;
                    break;
                }
                    
                if (crc((uint8_t *)ota_start_addr, len) != crc_value)
                    ota_ctrl[0] = OTA_STATUS_ERROR;
                else
                    ota_ctrl[0] = OTA_STATUS_OK;
            }
            break;
        case OTA_CTRL_READ_PAGE:
            if (ota_downloading)
                ota_ctrl[0] = OTA_STATUS_ERROR;
            else
            {
                ota_start_addr = *(uint32_t *)(buffer + 1);
                ota_ctrl[0] = OTA_STATUS_OK;
            }
            break;
        case OTA_CTRL_METADATA:
            if (OTA_STATUS_OK != ota_ctrl[0])
                break;
            if ((0 == ota_downloading) || (buffer_size < 1 + sizeof(ota_meta_t)))
            {
                const ota_meta_t  *meta = (const ota_meta_t *)(buffer + 1);
                int s = buffer_size - 1;
                if (crc((uint8_t *)&meta->entry, s - sizeof(meta->crc_value)) != meta->crc_value)
                {
                    ota_ctrl[0] = OTA_STATUS_ERROR;
                    break;
                }
                program_fota_metadata(meta->entry, 
                                      (s - sizeof(ota_meta_t)) / sizeof(meta->blocks[0]),
                                      meta->blocks);
            }
            else
            {
                ota_ctrl[0] = OTA_STATUS_ERROR;
            }
            break;
        case OTA_CTRL_REBOOT:
            if (OTA_STATUS_OK == ota_ctrl[0])
            {
                if (ota_downloading)
                    ota_ctrl[0] = OTA_STATUS_ERROR;
                else
                    platform_reset();
            }
            break;
        default:
            ota_ctrl[0] = OTA_STATUS_ERROR;
        }
    }
    else if (att_handle == ATT_OTA_HANDLE_DATA)
    {
        if (OTA_STATUS_OK == ota_ctrl[0])
        {
            if (   (buffer_size & 0x3) || (0 == ota_downloading)
                || (ota_page_offset + buffer_size > PAGE_SIZE))
            {
                ota_ctrl[0] = OTA_STATUS_ERROR;
                return 0;
            }
            
            memcpy(page_buffer + ota_page_offset,
                   buffer, buffer_size);
            ota_page_offset += buffer_size;
        }
    }
    else;

    return 0;
}

int ota_read_callback(uint16_t att_handle, uint16_t offset, uint8_t * buffer, uint16_t buffer_size)
{
    if (buffer == NULL)
    {
        if (att_handle == ATT_OTA_HANDLE_CTRL)
            return 1;
        else
            return 0;
    }

    if (att_handle == ATT_OTA_HANDLE_CTRL)
    {
        buffer[0] = ota_ctrl[0];
    }

    return buffer_size;
}

File: src/Tools/ota_service.nim
##
##  Copyright (C) INGCHIPS. All rights reserved.
##  This code is INGCHIPS proprietary and confidential.
##  Any use of the code for whatever purpose is subject to
##  specific written permission of INGCHIPS.
##

type
  prog_ver_t* {.importc: "prog_ver_t", header: "ota_service.h", bycopy.} = object
    major* {.importc: "major".}: cshort
    minor* {.importc: "minor".}: char
    patch* {.importc: "patch".}: char

  ota_ver_t* {.importc: "ota_ver_t", header: "ota_service.h", bycopy.} = object
    platform* {.importc: "platform".}: prog_ver_t
    app* {.importc: "app".}: prog_ver_t


const
  OTA_CTRL_START* = 0x000000AA
  OTA_CTRL_PAGE_BEGIN* = 0x000000B0
  OTA_CTRL_PAGE_END* = 0x000000B1
  OTA_CTRL_READ_PAGE* = 0x000000C0
  OTA_CTRL_SWITCH_APP* = 0x000000D0
  OTA_CTRL_REBOOT* = 0x000000FF
  OTA_STATUS_DISABLED* = 0
  OTA_STATUS_OK* = 1
  OTA_STATUS_ERROR* = 2

## *
## @brief Tell this module handles used by FOTA service
##
##    If FOTA related GATT service is *not* initialized by `ota_init_service`, then call this function.
##

proc ota_init_handles*(handler_ver: uint16; handle_ctrl: uint16; handle_data: uint16) {.
    importc: "ota_init_handles", header: "ota_service.h".}
proc ota_write_callback*(att_handle: uint16; transaction_mode: uint16;
                        offset: uint16; buffer: ptr uint8; buffer_size: uint16): cint {.
    importc: "ota_write_callback", header: "ota_service.h".}
proc ota_read_callback*(att_handle: uint16; offset: uint16; buffer: ptr uint8;
                       buffer_size: uint16): cint {.importc: "ota_read_callback",
    header: "ota_service.h".}